packages/story-editor/src/components/panels/design/layer/layer.js

```diff
@@ -350,6 +351,15 @@ function Layer({ element }) {
     })
   );
 
+  const { hasShapeMask, removeShapeMask } = useShapeMask();
```

Comment:
Hm, there's a potential performance issue here. Because `useShapeMask` depends on `selectedElements`, it will re-run every time selected elements are updated, which means every layer in the layer panel will re-render every time selection changes.

However, the list of selected elements is only used for the `shapeMaskElements` function in that hook, which isn't used at all inside the layer, so all the layers re-render for nothing.

I think you need to split this hook (`useShapeMask`) into two. Create a new hook that re-renders every time selection changes which defines which of the two selected elements is the shape, and which is the image. This new hook (maybe `useShapeMaskElements`) would only be used in the right-click menu (which won't be open while selection changes anyway). The function `addShapeMask` would probably go in this hook too.

Then this old hook is now "static" in terms of selection change, and can be used in both the layer panel and the right-click menu. And then you could actually refactor this `useShapeMask` hook to take the currently selected element as an argument, which it needs internally for a number of functions, and you now don't have to pass that anymore. It can even directly return the value of `hasShapeMask` as a boolean rather than as a function.

Reply:
Good call --- that cleans things up nicely especially not need to pass the element in multiple places.